cpp/benchmarks/ndsh/q09.cpp

```diff
@@ -111,46 +139,148 @@
   return amount;
 }
 
-void run_ndsh_q9(nvbench::state& state,
-                 std::unordered_map<std::string, cuio_source_sink_pair>& sources)
+[[nodiscard]] std::unique_ptr<cudf::column> compute_amount_transform(
+  cudf::column_view const& discount,
+  cudf::column_view const& extendedprice,
+  cudf::column_view const& supplycost,
+  cudf::column_view const& quantity,
+  rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+  rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref())
+{
+  CUDF_FUNC_RANGE();
+
+  std::string udf =
+    R"***(
+  void calculate_price(double * amount, double discount, double extended_price, double supply_cost, double quantity){
+    *amount = extended_price * (1 - discount) - supply_cost * quantity;
+  }
+  )***";
+
+  return cudf::transform({discount, extendedprice, supplycost, quantity},
+                         udf,
+                         cudf::data_type{cudf::type_id::FLOAT64},
+                         false,
+                         stream,
+                         mr);
+}
+
+[[nodiscard]] std::unique_ptr<cudf::column> compute_amount_ast(
+  cudf::column_view const& discount,
+  cudf::column_view const& extendedprice,
+  cudf::column_view const& supplycost,
+  cudf::column_view const& quantity,
+  rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+  rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref())
+{
+  CUDF_FUNC_RANGE();
+
+  cudf::ast::tree tree;
+  cudf::table_view table{std::vector{discount, extendedprice, supplycost, quantity}};
+
+  auto& discount_ref       = tree.push(cudf::ast::column_reference{0});
+  auto& extended_price_ref = tree.push(cudf::ast::column_reference{1});
+  auto& supplycost_ref     = tree.push(cudf::ast::column_reference{2});
+  auto& quantity_ref       = tree.push(cudf::ast::column_reference{3});
```

Comment:
I'm not familiar with these `push` function so please make sure not to create reference to temporary variables.

Reply:
they are references, they aren't temporaries. Yes, I was careful not to